cupy/cuda/cub.pyx

```diff
@@ -227,23 +242,45 @@ def can_use_device_segmented_reduce(int op, x_dtype, Py_ssize_t ndim, axis,
     return _cub_device_segmented_reduce_axis_compatible(axis, ndim)
 
 
-cdef _cub_reduce_dtype_compatible(x_dtype, int op, dtype=None,
-                                  bint segmented=False):
+cdef _cub_support_dtype(bint sum_mode, int dev_id):
+    if sum_mode:
+        support_dtype_dict = CUB_sum_support_dtype
+        with_half = CUB_sum_support_dtype_with_half
+        without_half = CUB_sum_support_dtype_without_half
+    else:
+        support_dtype_dict = CUB_support_dtype
+        with_half = CUB_support_dtype_with_half
+        without_half = CUB_support_dtype_without_half
+
+    if dev_id not in support_dtype_dict:
+        if int(device.get_compute_capability()) >= 53 and \
+                runtime.runtimeGetVersion() >= 9000:
+            support_dtype = with_half
+        else:
+            support_dtype = without_half
+
+        support_dtype_dict[dev_id] = support_dtype
+
+    return support_dtype_dict[dev_id]
+
+
+cdef _cub_reduce_dtype_compatible(x_dtype, int op, dtype=None):
+    dev_id = device.get_device_id()
```

Comment:
@y1r I just found that every `cupy.ndarray` has an attribute `device`:
```python
>>> import cupy as cp
>>> a = cp.arange(10)
>>> a.device
<CUDA Device 0>
>>> a.device.id
0
```
...and the underlying `MemoryPointer` also stores this information:
```python
>>> a.data
<cupy.cuda.memory.MemoryPointer object at 0x7f5b41c9e120>             
>>> a.data.device
<CUDA Device 0>
>>> a.data.device_id
0
```
Do you think it'd be easier to just use it? It'd involve changing a few function signatures from bottom up, so that the dev id can be passed from a higher level down. Could be a bit ugly.

I am not saying we should do this, just want to share this new finding. Should have known this earlier...😛